Refactor initialization and configuration to remove globals

tgamblin opened the pull request:
This is still WIP but it's almost done, so feel free to start reviewing.  It hurt a lot but it's needed. 😄

Spack's `__init__.py` has served us well but it's time to get rid of all the stuff in it.  It's essentially a small set of global variables to manage the state of a Spack run.  The fact that they are all in one top-level `__init__.py` has some serious negatives as Spack scales.  Specifically:

1. `spack` is loaded when `spack.anything` is imported.  This means everything in `__init__.py` has to be imported for every Spack run, which causes startup time to suffer (see #7585 and #7587).
2. All the imports mean that we have very little control over initialization order in Spack, and it's easy to get circular dependencies among the top-level modules. It also means we can't force certain imports to after before certain command-line options (see #6903 and #7692).
3. The imports in `__init__.py` pollute the top-level Spack namespace.  This will be dealt with in a future PR.

The way `spack.config` is currently written, it suffers from some of the same problems as (2), so it also needs to be made more lazy.

This PR will do the following:

- [x] Remove all module-scope initialization from `__init__.py`, so that pieces of spack can be imported selectively.
- [x] Rework `spack.config` so that module-scope state is in a `Configuration` class.  This makes it easy to override 
- [x] Simplify the `spack.config` interface to allow getting/setting either entire YAML files or specific parts (single variables).
- [x] Make it possible to handle command-line configuration settings and defaults through `spack.config`, by adding new, internal-only configuration scopes (i.e. scopes that don't correspond to files).  This unifies the way Spack handles user configuration settings and its own.
- [x] Replace all globals in `__init__.py` with configuration options via `spack.config`
- [x] Convert significant module-scope initialization in submodules into lazy initialization through functions or classes.

The last TODO was to make `__init__.py` load lazily for the `from spack import *` case used by `package.py` files, but I will continue this in another PR.

This should allow #6903 and #7692 to move forward and should hopefully help reduce the remaining startup latency discussed in #7587.  It should also make internal Spack code more clear, and give us more control going forward.  Finally, it moves us a little closer to #1385.

@alalazo @scheibelp @becker33 @mwkrentel @cyrush FYI

citibeth commented:
> by adding new, internal-only configuration scopes (i.e. scopes that don't correspond to files

In #2686, I made command-line configuration scopes.  This allowed one to stack multiple configuration scopes specified on the command line.  It's a feature I consider essential for environments, which you DON'T want the environment to rely on particular settings in the user's ~/.spack scope.

When I wrote #2686, I became aware of the complexity in ``__init.py``.  I am hopeful that this PR will reduce that complexity, and in general make re-implementing #2686 easier.  But I don't plan to re-implement as before; what I ACTUALLY need is for configuration scopes (as set up in files) to be added to Spack environments (#4585).  I will post more specifics there.

citibeth commented:
Until configuration scopes are implemented properly in #4585, the following quickie command-line utility would be REALLY useful: `scope-merge a b c` would read YAML files from a/, b/ and c/, merge them, and then write them to ~/spack.

citibeth commented:
I take it back on the config scopes.  I think that if we have appropriate config scope support in Spack Environments, we don't need it elsewhere.  And we are _almost_ there.

scheibelp commented on lib/spack/spack/operating_systems/cnl.py:
```diff
@@ -68,6 +69,9 @@ def find_compilers(self, *paths):
         return clist
 
     def find_compiler(self, cmp_cls, *paths):
+        # function-local so that cnl doesn't depend on spack.config
```
What happens if this depends on `spack.config`?

alalazo commented on lib/spack/spack/config.py:
```diff
@@ -84,23 +87,47 @@
+    ('site', os.path.join(spack.paths.etc_path, 'spack')),
+
+    # User configuration can override both spack defaults and site config
+    ('user', spack.paths.user_config_path)
```
Not sure we should do it in this PR, but I think it's worth swapping `user` scope with `site` scope, and make `site` the default for configuration. Based on my experience using multiple Spack versions at the same time, writing by default to `user` config is very likely to create conflicts if the configuration file format changed in the meanwhile.

tgamblin replied:
I think this is good but for another PR.

alalazo commented on lib/spack/spack/config.py:
```diff
@@ -84,23 +87,47 @@
+#: Hard-coded default values for some key configuration options.
+#: This ensures that Spack will still work even if config.yaml in
+#: the defaults scope is removed.
+config_defaults = {
```
:+1: to this

alalazo commented on lib/spack/spack/config.py:
```diff
@@ -206,61 +216,340 @@ def __repr__(self):
+        _validate_section(data, section_schemas[section])
+
+    def __repr__(self):
+        return '<InternalConfigScope: %s>' % self.name
```
Isn't `__repr__` usually supposed to return a string such that:
```python
b = eval(repr(a))
assert b == a
```
?

alalazo commented on lib/spack/spack/config.py:
```diff
@@ -206,61 +216,340 @@ def __repr__(self):
-ConfigScope('site/%s' % _platform, os.path.join(_site_path, _platform))
+        If ``scope`` is ``None`` or not provided, return the merged contents
+        of all of Spack's configuration scopes.  If ``scope`` is provided,
+        return only the confiugration as specified in that scope.
```
typo: confiugration

alalazo commented on lib/spack/spack/config.py:
```diff
@@ -206,61 +216,340 @@ def __repr__(self):
+
+        We use ``:`` as the separator, like YAML objects.
+
+        TODO: Currently only handles maps. Think about lists if neded.
```
Do you mean lists of objects? Something like:
```
spack.config.get('config:tests:0:url')
```
to get the `url` key from the item at index 0 of a list of things that looks like:
```python
[{ url: 'foo' },
 { url: 'bar'}]
```
? Or are you referring to another behavior that is not handled?

tgamblin replied:
I mean something like that, but I think we'd need to differentiate the syntax, e.g.:

```python
spack.config.get('config:tests[0]:url')
```

The keys following `:` are assumed to be strings, and result in `index[key]` in the Python code.  The indexes for lists are integers, and you'd need to convert them before indexing.  Yes, we could do that by looking at the type of structure we're indexing, but I think the above is clearer anyway because you cana clearly see what's expected to be a list and what's expected to be a dict in the path specifier.

I'm basically just saying we're not doing anything to address the list case currently, so you can't ask for, e.g.:

```python
spack.config.get('repos[2]')
```

scheibelp replied:
This function is used frequently so IMO the `TODO` should be a `#` comment vs. living in the docstring.

alalazo commented on lib/spack/spack/config.py:
```diff
@@ -206,61 +216,340 @@ def __repr__(self):
+            value = value.get(key, default)
+        return value
+
+    def set(self, path, value, scope=None):
```
Don't know how much this would be a problem in reality, but I think the semantic of this class is kind of confusing. If I read the code correctly,  by default when we `get` a value we retrieve a merged view of all the scopes. Instead, when we `set` a value, we set it in a particular scope.

Am I wrong saying that this means that:
```python
# We set the value in the highest precedence scope by default
spack.config.set(somepath, a) 

# We merge all the scopes by default
b = spack.config.get(somepath)

# This might be False, e.g. if 'somepath' is a dict spread over scopes
assert a == b
```
might happen? In case is this the behavior we want?

tgamblin replied:
This would never be false, unless you specified a scope in the first `spack.config.set`.  FWIW, this new syntax doesn't actually change the semantics of working with the config system.  It just makes it easier to partially update something without getting the entire dictionary for a level.

alalazo replied:
Say we have two scopes for `foo.yaml`:
```yaml
# first scope
foo:
   bar: 5
# second scope
foo:
    baz: 6
```
Am I wrong on the following snippet:
```python
spack.config.set('foo', {'foobar': 6}) 
b = spack.config.get('foo')
assert b == {'foobar': 6, 'baz': 6}
```
?

alalazo commented on lib/spack/spack/config.py:
```diff
@@ -206,61 +216,340 @@ def __repr__(self):
+
+
+@contextmanager
+def override(path, value):
```
Nice addition!

tgamblin commented:
@scheibelp @alalazo: this is ready for another review.  It should be mostly done now.

Note that I will be submitting some other PRs based on this one.  I think this PR has gotten large enough with just the global removal that it should be split up.  Probably best to look at each commit if you review this, as there are a lot of changes.

citibeth commented:
@scheibelp @tgamblin Now that we have (very soon) Spack Environments --- with a lot of scoping features built-in --- I don't think we need command line scopes. At least I have no use case for them.  But certainly, any scoping stuff here needs to work with Environment scopes (of which there might be multiple; see `env.yaml` file).

tgamblin commented on lib/spack/docs/conf.py:
```diff
@@ -192,9 +192,9 @@ def setup(sphinx):
 #
 # The short X.Y version.
-version = str(spack.spack_version.up_to(2))
+version = '.'.join(str(s) for s in spack.spack_version_info[:2])
```
The main reason was to remove any dependency on `spack` submodules in `spack/__init__.py`.  Also, it was never used as a version. I agree that a `Version` is nicer, but the whole point of the PR is to reduce interdependencies within Spack.  So I'd like to keep this one.

tgamblin replied:
Maybe this is clearer: if we revert this, importing any spack submodule implicitly imports `spack.version`, because it is in `spack/__init__.py`.

tgamblin commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -412,7 +414,7 @@ def relocate_package(workdir, allow_root):
     """
     buildinfo = read_buildinfo_file(workdir)
-    new_path = spack.store.layout.root
+    new_path = spack.store.store().layout.root
```
This would require the `spack` module to have a symbol called `store` and to know about `spack.store`.  So I'm trying to move away from that.  I do agree this could be nicer.  I'm not 100% sure what the best structure is, though.

I would consider the singletons introduced here to be a stopgap -- they allow lazy loading but they're still essentially global.  I'd like to eventually get to the point where we can instantiate a `Spack` object that just has a `store`, and even have multiple of them in the same process, but we're not there yet.  So I'm not wed to this, but it speeds up init for now.

scheibelp replied:
After thinking on it more I'm not even sure there's a problem with this, and also agree that my suggestion may not be the way to handle it if it was a problem (just noting this for future reference).

tgamblin commented on lib/spack/spack/cmd/bootstrap.py:
```diff
@@ -86,5 +79,5 @@ def bootstrap(parser, args, **kwargs):
                     (spec_to_install, requirement))
             kwargs['explicit'] = True
-            package = spack.repo.get(spec_to_install)
+            package = spack.repo.path().get(spec_to_install)
```
This one I'd be open to changing in a separate PR if people think it's worthwhile, but we should think about it -- I would like to move away from global state where possible, and a module-scope `get()` function implies that there is a singleton somewhere.

tgamblin replied:
I added convenience functions for `get` and `all_package_names` to address this.

scheibelp replied:
@tgamblin I meant that I think `spack.repo.path()` should be renamed to `spack.repo.get()`

If that was clear then regarding:

> a module-scope get() function implies that there is a singleton somewhere.

That's what it is though isn't it? The docstring for `path` reads:

> Get the singleton RepoPath instance for Spack.

tgamblin replied:
Yes, there is a singleton now.  I just don't want to have to wrap the API for singletons at the module level and maintain it.

In the changes I committed, I kept `path` as the singleton, and it's used in several places.  But I made top-level convenience functions for `get` and `all_package_names` that just call `path().get()` and `path().all_package_names()`.  Two wrapper functions isn't the worst.  But my main point is that at some point I would like *not* to have singletons at all.

scheibelp replied:
> there is a singleton now. I just don't want to have to wrap the API for singletons at the module level and maintain it.

@tgamblin I'm confused. I'm only suggesting changing the name of the function. I'm not sure how that could create additional maintenance problems.

tgamblin commented on lib/spack/spack/compilers/__init__.py:
```diff
@@ -337,17 +339,18 @@ def compiler_for_spec(compiler_spec, arch_spec):
-    config_scopes = spack.config.config_scopes
-    scope_to_compilers = dict()
-    for scope in config_scopes:
+    config = spack.config.get_configuration()
```
I like this one.

tgamblin replied:
I added some convenience functions to simplify this.  The code does read better now.

tgamblin commented on lib/spack/spack/package.py:
```diff
@@ -1591,6 +1604,14 @@ def build_process():
+    def test_check(self):
+        """Tests override this function to do assertions in do_install().
+
+        Normally returns ``True``, but subclasses should return ``None``
```
I tweaked the wording on this and updated the docs.

tgamblin commented:
@scheibelp: I think I've addressed most of the comments.  If this looks good to you can you rebase/merge?

scheibelp commented on lib/spack/spack/config.py:
```diff
@@ -207,61 +208,345 @@ def __repr__(self):
+
+        # we make a special scope for spack commands so that they can
+        # override configuration options.
+        _configuration.push_scope(InternalConfigScope('command_line'))
```
@tgamblin to clarify I mean to say that any PR which wants to introduce custom configuration scopes will likely have to change the way this is handled (at which point they will have to add a unit test for the alternative handling for command line scopes). Not necessarily a problem, but FYI  (e.g. this would have to be resolved for environment support)

tgamblin replied:
Thanks -- I think this is ok

alalazo commented on lib/spack/spack/config.py:
```diff
@@ -207,61 +208,349 @@ def __repr__(self):
+
+def scopes():
+    """Convenience function to get list of configuration scopes."""
+    return config().scopes
```
Why do we need wrappers to mask the singleton? Can't we avoid this?

tgamblin replied:
It's not easy to put property semantics at module scope.  You can, but IMHO it's not worth the effort.

scheibelp replied:
My preference is consistency. Earlier we had

`config.get('packages:...')`

and 

`config.get_configuration().scopes()`

(`get_configuration()` has since been renamed to `config()`). Having two different ways of interacting with the singleton was confusing IMO. I think masking the singleton is OK as long as it's clear what is going on when someone looks at `config.py` specifically.

alalazo commented on lib/spack/spack/modules/common.py:
```diff
@@ -56,22 +56,23 @@
 
-#: Root folders where the various module files should be written
-roots = spack.config.get_config('config').get('module_roots', {})
+#: config section for this flie
```
Typo: file

alalazo commented on lib/spack/spack/package.py:
```diff
@@ -1552,10 +1560,15 @@ def build_process():
+        # do any checks needed by the test suite
+        # return without building if only testing
+        if not self.unit_test_check():
+            return
```
I am not sure I get the logic of this part. Is this used only *within* unit tests to disable building?

alalazo commented on lib/spack/spack/spec.py:
```diff
@@ -2166,16 +2173,23 @@ def _normalize_helper(self, visited, spec_deps, provider_index):
+                if dep:
+                    merge = (
+                        # caller requested test dependencies
+                        tests is True or (tests and self.name in tests) or
```
>  If we refactor this as we've discussed in the past, so that repo gets the class and not an instance,

What do you mean by this?

tgamblin commented on lib/spack/spack/__init__.py:
```diff
@@ -23,159 +23,18 @@
 ##############################################################################
-import multiprocessing
-import os
 import sys
```
nope. fixed.

tgamblin commented on lib/spack/spack/caches.py:
```diff
@@ -0,0 +1,71 @@
+_fetch_cache = None
+
+
+def misc_cache():
```
I like it!

tgamblin commented on lib/spack/spack/concretize.py:
```diff
@@ -36,29 +36,60 @@
 
+#: controls whether we check that compiler versions actually exist during
+#: concretization.  Used for testing.
+check_for_compiler_existence = True
```
fixed

tgamblin commented:
@alalazo @scheibelp: I've pushed some updates based on your comments, including adding @alalazo's singleton (lazyobject) and updating some comments that weren't clear.  I *think* this is ready to go.

The only thing I haven't done is update the `config` syntax.  I think it would be good to support lists and maybe `::` overrides in `set`, and also it might be good to make it so you can `spack config get` these paths on the command line, but I think that's not needed for this PR.  I'd rather commit something focused on that that we can review independently.

Note: I reworded the way `database` and `install_mockery` work based on the refactor of `spack.store`.  I think they are a lot simpler now.  I also removed some placed where these fixtures yielded custom objects as `namedtuple`s.  It seems like that was being used to expose both the original and the mock version of the fixture, but the real version was never used, and the tests would all call `database.mock`.  I simplified this by just having the fixture yield the mock database.

tgamblin commented:
@scheibelp: if you think this is good i think you can rebase and merge.
